File: 2 Mediam/0056 Merge Intervals/Solution.py
from typing import List

class Solution:
    # Solution 1: Brute Force
    # O(nlog(n)), O(1)
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals
        
        i = 1
        intervals.sort()
        while i < len(intervals):
            start, end = intervals[i-1][0], intervals[i-1][1]
            if intervals[i][0] <= end:
                intervals[i] = [min(intervals[i][0], start), max(intervals[i][1], end)]
                intervals[i-1] = [-1,-1]
            i += 1
        return [i for i in intervals if i != [-1,-1]]
        # return filter(lambda x: x!=[-1,-1], intervals)

File: 2 Mediam/0056 Merge Intervals/Solution.java
import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> res = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        res.add(intervals[0]);

        for (int i = 1; i < intervals.length; i++) {
            if (i < intervals.length && res.get(res.size() - 1)[1] >= intervals[i][0]) {
                int start = res.get(res.size() - 1)[0];
                int end = Math.max(res.get(res.size() - 1)[1], intervals[i][1]);
                res.remove(res.size() - 1);
                res.add(new int[] { start, end });
            } else {
                res.add(intervals[i]);
            }
        }

        int[][] arr = new int[res.size()][2];
        for (int i = 0; i < res.size(); i++) {
            arr[i] = res.get(i);
        }
        return arr;
    }
}